File: 101.symmetric-tree.py
#
# @lc app=leetcode id=101 lang=python3
#
# [101] Symmetric Tree
#
from typing import List
# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        # recursively
        # if not root:
        #     return True
        # return self.compare_recursive(root.left, root.right)

        # iteratively
        if not root:
            return True
        stack = [root.left, root.right]

        while len(stack) > 0:
            node1 = stack.pop()
            node2 = stack.pop()
            if not node1 and not node2:
                continue
            if not node1 or not node2:
                return False
            
            if node1.val == node2.val:
                stack.append(node1.left)
                stack.append(node2.right)
                
                stack.append(node1.right)
                stack.append(node2.left)
            else:
                return False
        return True

    def compare_recursive(self, node1: TreeNode, node2: TreeNode) -> bool:
        if not node1 and not node2:
            # 둘 다 None인 경우
            return True
            
        if not node1 or not node2:
            # 둘 다 None인 경우를 제외하고, 둘 중 하나만 None이면
            return False

        if node1.val == node2.val:
            return self.compare(node1.left, node2.right) and self.compare(node2.left, node1.right)
        else:
            return False

# @lc code=end
'''
[1,2,2,null,3,null,3]
'''

File: 101.symmetric-tree.js
/*
 * @lc app=leetcode id=101 lang=javascript
 *
 * [101] Symmetric Tree
 */

// Definition for a binary tree node.
class TreeNode {
  constructor(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
  }
}
// @lc code=start
/**
 * @param {array} stack
 * @return {array}
 */
const search = (node) => {
  if (!node) {
    return stack;
  }

  const stack = [node.left, node.right];
  let node1;
  let node2;

  while (stack.length > 0) {
    node1 = stack.pop();
    node2 = stack.pop();

    if (node1 === null && node2 === null) {
      // 노드의 끝까지 탐색한 경우 (둘다 null)
      continue;
    }

    if (node1 === null || node2 === null) {
      // 위의 조건문을 통과하고, 여기서 true가 나오면 한 쪽만 null인것
      return false;
    }
    
    if (node1.val !== node2.val) {
      // 두 값이 다른 경우 
      return false;
    }

    stack.push(node1.left);
    stack.push(node2.right);
    stack.push(node1.right);
    stack.push(node2.left);
    // 대칭을 찾는 것이니 좌-우 적절히 순서대로 넣어서
    // 나중에 pop()으로 비교할 수 있게 함
  }

  // 반복문을 멀쩡히 나오면 대칭인 것
  return true;
};
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const isSymmetric = (root) => {
  return search(root);

};
// @lc code=end
